New Design for Visualize

stormpython opened the pull request:
Changes to the appearance of the visualize interface.

stormpython commented:
jenkins test it

rashidkpc commented on src/plugins/kibana/public/visualize/styles/main.less:
```diff
@@ -2,21 +2,179 @@
 @import (reference) "~bootstrap/less/list-group";
 @import (reference) "~ui/styles/list-group-menu";
 
```
Move colors out to variables, reuse `kibanaGrays`

rashidkpc commented on src/plugins/kibana/public/visualize/styles/main.less:
```diff
@@ -2,21 +2,179 @@
+
+  span {
+    padding: 5px 0px;
+    background-color: #FFFFFF;
```
`@white`

rashidkpc commented on src/ui/public/styles/pagination.less:
```diff
@@ -29,6 +29,8 @@ paginate {
 
           a {
             text-decoration: none;
+            background-color: #ffffff;
```
`@white`

rashidkpc commented on src/ui/public/partials/saved_object_finder.html:
```diff
@@ -1,47 +1,65 @@
-  <ul
-    class="list-group list-group-menu"
+<paginate list="finder.hits" per-page="20" class="paginate-style">
+  <table class="list-group list-group-menu table table-striped"
```
Why change this to a table?

rashidkpc commented:
Needs a merge with `feature/design`.

rashidkpc commented:
Saved object finder needs some polishing. Padding, etc.

**@alt74's design**
![screen shot 2016-02-11 at 6 05 55 pm](https://cloud.githubusercontent.com/assets/1250387/12996152/7911aea6-d0ea-11e5-9c42-15965c3ec01b.png)

**Current**
![screen shot 2016-02-11 at 6 06 15 pm](https://cloud.githubusercontent.com/assets/1250387/12996153/7cf20a98-d0ea-11e5-942b-146619b59d0b.png)

stormpython commented:
@rashidkpc I can change paginate back to a list and merge in feature/design. Can you point me to the link for @alt74's design for the saved object finder because that looks different from the pics I've been looking at.

stormpython commented:
@rashidkpc this is ready for review again.

rashidkpc commented:
It seems like what you're going for in the `capitalize` directive is title-case, makes the naming sort of confusing. Further, its english specific. Can we just use `text-transform: capitalize`? Sure there's a few words that will be capitalized when they wouldn't be on the cover of a book, but I don't think its a big deal.

rashidkpc commented on src/plugins/kibana/public/visualize/styles/main.less:
```diff
@@ -2,21 +2,247 @@
 @import (reference) "~bootstrap/less/list-group";
 @import (reference) "~ui/styles/list-group-menu";
 
+@icon-background-color: #416376;
```
Whats the deal with this color? Do we use it anywhere else? If so, let's use a variable, if not, let's find a color we do use elsewhere and use its variable

rashidkpc replied:
Also, none of these styles are scoped in any way, do we really need special styling for the saved object loader just for this page? Can you work with @alt74 to find a style that works both in pages and in the drop downs?

rashidkpc commented on src/plugins/kibana/public/visualize/styles/main.less:
```diff
@@ -2,21 +2,247 @@
+  background-color: @kibanaGray6;
+}
+
+.form-text {
```
Is form-text blue anywhere else, is there a way we can apply this more generically?

rashidkpc commented on src/plugins/kibana/public/visualize/styles/main.less:
```diff
@@ -2,21 +2,247 @@
+  color: @kibanaBlue2;
+}
+
+.link-color {
```
same deal here

rashidkpc commented on src/plugins/kibana/public/visualize/styles/main.less:
```diff
@@ -2,21 +2,247 @@
+
+/* saved_object_finder.html striped table colors */
+.table-striped > tbody > tr:nth-of-type(even) {
+  background-color: @kibanaGray6;
```
Is there something more semantic we can use as the naming for all of the table styles that use `kibanaGray6`? Something they have in common?

rashidkpc commented on src/ui/public/partials/saved_object_finder.html:
```diff
@@ -1,42 +1,70 @@
     ng-class="{'select-mode': finder.selector.enabled}">
 
+    <li class="list-header list-group-item list-group-menu-item">
+      <ul class="li-not-striped">
```
this feels off. Can we create a style for the paginate directives and apply it, instead of making classes that aren't reused outside of paginate?

rashidkpc commented on src/ui/public/styles/pagination.less:
```diff
@@ -29,6 +29,9 @@ paginate {
 
           a {
             text-decoration: none;
+            background-color: @white;
```
This applies to all paginate directives, did you mean to do this?

stormpython replied:
Yes, according to the spec, the pagination style is consistent through the different views.

rashidkpc commented on src/ui/public/styles/variables/for-theme.less:
```diff
@@ -217,7 +217,7 @@
 
 // Paginate ====================================================================
-@paginate-page-link-active-color: @text-color;
+@paginate-page-link-active-color: @kibanaGray2;
```
Same here, did you mean to change this everywhere?

stormpython replied:
Yes.

stormpython commented:
@panda01, @rashidkpc I refactored the less files. It turns out, there were saved object finder specific styles in `base.less`, which is where I moved styles that affected the saved object finder directive. I also refactored the `wizard` css. Finally, I removed the `capitalize` filter in favor of the css `text-transform` property.

rashidkpc commented on src/plugins/kibana/public/visualize/styles/main.less:
```diff
@@ -3,44 +3,98 @@
+  padding-left: 0;
+  padding-right: 0;
+
+  @media (min-width: 1280px) {
```
Where did this 1280px come from? `@screen-lg` is 1200px?

rashidkpc commented on src/plugins/kibana/public/visualize/wizard/wizard.js:
```diff
@@ -7,8 +7,8 @@ import RegistryVisTypesProvider from 'ui/registry/vis_types';
 
-const templateStep = function (num, txt) {
-  return '<div ng-controller="VisualizeWizardStep' + num + '" class="container vis-wizard">' + txt + '</div>';
+var templateStep = function (num, txt) {
```
Change this back to const

rashidkpc commented on src/ui/public/paginated_table/paginated_table.html:
```diff
@@ -33,5 +33,5 @@
   <!-- <paginate-controls></paginate-controls> -->
   <div class="pagination-container" ng-transclude></div>
-
+  
```
Stray white space, you might configure your editor to strip trailing white space on save

rashidkpc commented on src/plugins/kibana/public/visualize/wizard/step_2.html:
```diff
@@ -1,37 +1,26 @@
-      <select
-              class="form-control"
-              ng-model="indexPattern.selection"
-              ng-options="pattern as pattern for pattern in indexPattern.list | orderBy:'toString()'">
```
Why was the orderBy removed? The index pattern list is no longer in alphabetical order

rashidkpc commented:
![screen shot 2016-03-10 at 2 41 58 pm](https://cloud.githubusercontent.com/assets/1250387/13685580/561eabf2-e6ce-11e5-93ba-00d51ec39715.png)

White space should be the same height as the ping logo

rashidkpc commented:
Reviewing @alt74's design, step 2 doesn't match:

**This pull**
![screen shot 2016-03-10 at 3 00 41 pm](https://cloud.githubusercontent.com/assets/1250387/13686062/1033ba44-e6d1-11e5-8052-b575a1aebe85.png)

**The design**
![screen shot 2016-03-10 at 3 00 02 pm](https://cloud.githubusercontent.com/assets/1250387/13686067/1651644e-e6d1-11e5-83cc-ace56c1c448b.png)

![screen shot 2016-03-10 at 3 00 17 pm](https://cloud.githubusercontent.com/assets/1250387/13686071/197d8166-e6d1-11e5-9fea-080d682253eb.png)

rashidkpc commented:
Some padding issues on the right:

![screen shot 2016-03-10 at 3 04 34 pm](https://cloud.githubusercontent.com/assets/1250387/13686111/6e70d92a-e6d1-11e5-9ef5-a52cc1f44813.png)

rashidkpc commented on src/plugins/kibana/public/visualize/wizard/wizard.js:
```diff
@@ -56,16 +56,8 @@ module.controller('VisualizeWizardStep2', function ($route, $scope, $location, t
-  });
-
-  $scope.$watch('indexPattern.selection', function (pattern) {
+  $scope.goToUrl = function (pattern) {
```
I don't think this is required, the index pattern selector is a drop down in the design. Though for future reference, for stuff like this you can use kbn-href

rashidkpc commented:
Sounds good, @alt74 can you update the mockups?

rashidkpc commented on src/plugins/kibana/public/visualize/wizard/step_2.html:
```diff
@@ -1,37 +1,26 @@
+      </div>
+      <ul class="striped list-group">
+        <li class="list-group-item" ng-repeat="pattern in indexPattern.list">
+          <a class="index-link" href ng-click="goToUrl(pattern)">{{pattern}}</a>
```
Don't use ng-click, use kbn-href

rashidkpc commented:
One of these is hovered, the other isn't, we need a more obvious hover effect here. @alt74 ?

![screen shot 2016-03-10 at 3 20 17 pm](https://cloud.githubusercontent.com/assets/1250387/13686591/affb09f4-e6d3-11e5-8a6d-9cf46b419703.png)

rashidkpc commented on src/plugins/kibana/public/visualize/wizard/step_2.html:
```diff
@@ -1,37 +1,26 @@
+        <div class="panel-heading">Index Patterns</div>
+      </div>
+      <ul class="striped list-group">
+        <li class="list-group-item" ng-repeat="pattern in indexPattern.list">
```
This list should be paginated like the saved object loader.

Also, in all of the other columns, the entire row is clickable. In this one, only the name of the pattern is clickable.

rashidkpc commented on src/ui/public/directives/saved_object_finder.js:
```diff
@@ -49,6 +49,38 @@ module.directive('savedObjectFinder', function ($location, $injector, kbnUrl, Pr
+        self.isAscending = !self.isAscending;
+
+        if (hits.length) {
+          if (self.isAscending) {
```
replace all of this with `return self.isAscending ? _.sortBy(hits, 'title') : _.sortBy(hits, 'title').reverse();`

rashidkpc commented on src/ui/public/partials/saved_object_finder.html:
```diff
@@ -1,27 +1,50 @@
+          ng-attr-placeholder="{{finder.properties.nouns}} Filter..."
+          ng-keydown="finder.filterKeyDown($event)"
+          class="form-control"
+          style="text-transform: capitalize"
```
Get rid of the inline style here

rashidkpc commented on src/ui/public/partials/saved_object_finder.html:
```diff
@@ -1,27 +1,50 @@
 
+    <li class="list-group-item list-group-menu-item">
+      <ul>
+        <li class="finder-type paginate-heading" ng-show="finder.properties.noun === 'Visualization'">Type</li>
```
I don't like the idea of putting a type specific check in the saved object loader. I'd rather take out this column than do this

rashidkpc commented on src/ui/public/partials/saved_object_finder.html:
```diff
@@ -29,19 +52,26 @@
+          ng-blur="finder.hitBlur($event)"
+          ng-click="finder.preventClick($event)">
+        <ul>
+          <li ng-show="finder.properties.noun === 'Visualization'" class="finder-type">
```
Same here, why not just look to see if there's a hit.icon on the hit like the row below?

rashidkpc commented on src/ui/public/styles/base.less:
```diff
@@ -289,14 +289,102 @@ table {
+    }
+  }
+
+  #finder-hit-count, #finder-manage-object {
```
Why IDs instead of classes?

rashidkpc replied:
Also, these styles are specific to the saved object finder, put them under `saved-object-finder`

rashidkpc commented on src/ui/public/styles/base.less:
```diff
@@ -289,14 +289,102 @@ table {
+      }
+
+      li:nth-child(odd) {
+        background-color: transparent;
```
Given these are both transparent do we need both a rule for odd and one for even? Couldn't they be a single rule?

rashidkpc commented on src/ui/public/partials/saved_object_finder.html:
```diff
@@ -29,19 +52,26 @@
+      <a ng-href="{{finder.makeUrl(hit)}}"
+          ng-blur="finder.hitBlur($event)"
+          ng-click="finder.preventClick($event)">
+        <ul>
```
A list inside the link? Just for the icon?

rashidkpc commented on src/ui/public/partials/saved_object_finder.html:
```diff
@@ -1,27 +1,50 @@
     ng-class="{'select-mode': finder.selector.enabled}">
 
+    <li class="list-group-item list-group-menu-item">
+      <ul>
```
This ul inside another ul doesn't feel right. It sort of feels like you're going for a table, but I don't think we need a table here either?

rashidkpc commented on src/ui/public/partials/saved_object_finder.html:
```diff
@@ -1,27 +1,50 @@
-      class="form-control"
-      name="filter"
-      type="text"
-      autocomplete="off" />
```
Why was this removed?

rashidkpc commented:
@panda01 Can you also look at this, see if I missed anything?

stormpython commented:
Yeah, I caught that. Still working on more changes.

rashidkpc commented on src/ui/public/directives/saved_object_finder.js:
```diff
@@ -49,6 +49,22 @@ module.directive('savedObjectFinder', function ($location, $injector, kbnUrl, Pr
+       * @return {Array} Array sorted either ascending or descending
+       */
+      self.sortHits = function (hits) {
+        return self.isAscending ? _.sortBy(hits, 'title') : _.sortBy(hits, 'title').reverse();
```
this should be assigning `self.hits` to the result of the ternary instead of returning it.

rashidkpc commented on src/ui/public/partials/saved_object_finder.html:
```diff
@@ -1,27 +1,49 @@
+          <i
+            class="fa"
+            ng-click="finder.sortHits(finder.hits)"
+            ng-class="{
```
replace this with the ternary form:

`finder.isAscending ? 'fa-caret-up' : 'fa-caret-down'`

rashidkpc commented on src/plugins/kibana/public/visualize/wizard/step_2.html:
```diff
@@ -1,37 +1,26 @@
+      </div>
+      <ul class="striped list-group">
+        <li class="list-group-item" ng-repeat="pattern in indexPattern.list | orderBy: 'toString()'">
+          <a class="index-link" href ng-click="goToUrl(pattern)">{{pattern}}</a>
```
Use a `kbn-href` here instead of ng-click. ng-click should be avoided for navigation because it breaks open-in-new-tab functionality.

rashidkpc commented:
Polished design via @alt74 for step 2. We should implement that index pattern selection as a reusable directive, paginated, selectable list. There's lots of places that could be reused. Heck, we could probably reuse it in the future to simplify the saved object loader.

The directive should have an isolate scope that takes 3 parameters:

```
scope: {
  perPage: '=', // The number of items to show on each page
  list: '=', // An array of objects
  onSelect: '=', // a function to pass the selected object to
}
```

The objects in the list should either require a title property, or you could make that property configurable

<img width="1273" alt="screen shot 2016-03-11 at 10 15 41 am" src="https://cloud.githubusercontent.com/assets/1250387/13709652/58329c06-e772-11e5-8ef9-8cae90d09e9f.png">

stormpython commented:
There is a bug in the kbn-href directive. The pull request that corrects it is #6517. This will need to be merged into this pr.

rashidkpc commented:
![screen shot 2016-03-14 at 1 19 13 pm](https://cloud.githubusercontent.com/assets/1250387/13758637/a01225d8-e9e7-11e5-9b67-c1a066c0b1aa.png)

rashidkpc commented:
![screen shot 2016-03-14 at 1 23 04 pm](https://cloud.githubusercontent.com/assets/1250387/13758696/e7743c04-e9e7-11e5-9926-da2a5a440725.png)

There's a hover effect on the active tab, but not on the inactive ones. This should be the other way around

rashidkpc commented on src/ui/public/styles/base.less:
```diff
@@ -289,28 +289,108 @@ table {
+    }
+  }
+
+  .finder-hit-count, .finder-manage-object {
```
Put this under the saved object finder above. Same with all of the .finder* classes

stormpython replied:
@rashidkpc, I am confused. Aren't they already under the `saved-object-finder`?

rashidkpc replied:
So they are, my bad

rashidkpc commented:
This LGTM now, merging
